### PR TITLE
Reduce GPU device object

### PR DESCRIPTION
**Pull request by okuta**
This PR reduce GPU device object. In most cases it is less expensive to keep the GPU device id.

No compatible point.
- `ndarray.device` will not be `None`. So far it is `None` when the array size is 0.

### CODE REVIEW

**Comment by niboshi** [comment]
This PR fixes https://github.com/chainer/chainer/issues/3702.

**Comment by niboshi** [comment]
Please resolve conflict.

**Comment by okuta** [comment]
I rebased.

**Comment by niboshi** [comment]
>No compatible point.
>- `ndarray.device` will not be `None`. So far it is `None` when the array size is 0.

I think this can be considered compatibility-preserving, as this behavior is not documented, and the change is toward generalization rather than specialization.

How do you think?

**Comment by okuta** [comment]
OK, I agree your opinion.
It is assumed that many users have not written code based on 0-size array returning `None` device.

**Comment by okuta** [comment]
@niboshi Cloud you check this PR?

File: cupy/core/core.pyx [hunk 1]
```diff
@@ -377,17 +377,20 @@ cdef class ndarray:
         if self.size == 0:
             return self.astype(self.dtype, copy=True, order=order)
 
-        if (self.data.device is None or
-                self.data.device.id == device.get_device_id()):
+        dev_id = device.get_device_id()
+        if self.data.device_id == device.get_device_id():
```
Comment (niboshi) [inline cupy/core/core.pyx]: rhs should be `dev_id`?

**Comment by niboshi** [comment]
@okuta 
Sorry, could you check my review?

File: cupy/core/core.pyx (hunk 1 above)
Comment (okuta) [reply cupy/core/core.pyx]: Thanks, I fixed.

**Comment by niboshi** [comment]
jenkins, test this please

**Comment by okuta** [comment]
jenkins, test this please.

**Comment by niboshi** [comment]
Considering the fact this change affects Chainer's behavior (chainer/chainer#5071), I've come to think this PR should be `non-compat`. How do you think? @okuta

**Comment by okuta** [comment]
Hmm, It is not documented behavior. And, Almost user do not assume `None` when array size is zero.

**Comment by niboshi** [comment]
OK, maybe we can leave without `no-compat` in CuPy, and make `no-compat` in Chainer side.

**Comment by okuta** [comment]
OK, I understood.

**Comment by niboshi** [comment]
Jenkins, test this please

**Comment by niboshi** [comment]
Jenkins, test this please

**Comment by niboshi** [comment]
Jenkins, test this please

**Comment by niboshi** [comment]
Jenkins, test this please

**Comment by niboshi** [comment]
LGTM

**Comment by okuta** [comment]
Thanks

**Pull request merged**